flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -212,6 +212,13 @@ private Runnable restartTasks(final Set<ExecutionVertexVersion> executionVertexV
 
 			resetForNewExecutionIfInTerminalState(verticesToRestart);
 
+			try {
+				restoreState(verticesToRestart);
```

Comment:
Does it make sense to move this block into `resetForNewExecutionIfInTerminalState()`?

Reply:
I think both work. But I moved it out because the name `resetForNewExecutionIfInTerminalState`  is a bit deceiving, since we are restoring state for all verticesToRestart , not only the vertices in terminal state.

But I'm thinking whether we can change `resetForNewExecutionIfInTerminalState`. Rather than reset vertices in terminal state, we should check all these vertices are in terminal state and then do the reset. Checking failures may indicate a bug.

Reply:
It would be better to check if all vertices are in terminal state.

Reply:
done.
I'd like to keep the restore block directly in `DefaultScheduler#restartTasks` so that we can return to skip schedulingStrategy.restartTasks is an error happen in restore. Otherwise we need to change `resetForNewExecutions` to have an return value.

Reply:
ok